drivers/serial/serial_test.c

```diff
@@ -9,18 +9,100 @@
  * devices for the "vnd,serial" devicetree compatible used in test code.
  */
 
+#include <zephyr/device.h>
 #include <zephyr/drivers/uart.h>
+#include <zephyr/drivers/uart/serial_test.h>
+#include <zephyr/sys/ring_buffer.h>
 #include <zephyr/zephyr.h>
 
 #define DT_DRV_COMPAT vnd_serial
+struct serial_vnd_data {
+	struct ring_buf *written;
+	struct ring_buf *read_queue;
+	serial_vnd_write_cb_t callback;
+	void *callback_data;
+};
 
 static int serial_vnd_poll_in(const struct device *dev, unsigned char *c)
 {
-	return -ENOTSUP;
+	struct serial_vnd_data *data = dev->data;
+	uint32_t bytes_read;
+
+	if (!data || !data->read_queue) {
```

Comment:
FYI, some MISRA rules require these kind of tests to be spelled out, e.g. `(data == NULL) || (data->read_queue == NULL)`.

Ditto for similar statements below.